wagtail/contrib/redirects/signal_handlers.py

```diff
@@ -0,0 +1,95 @@
+import logging
+
+from typing import Optional
+
+from django.conf import settings
+from django.db.models import Q
+
+from wagtail.core.models import BaseLogEntry, Page, PageBase
+from wagtail.core.utils import BatchCreator, get_dummy_request
+
+from .models import Redirect
+
+
+logger = logging.getLogger(__name__)
+
+
+class BatchRedirectCreator(BatchCreator):
+    """
+    A specialized ``BatchCreator`` class for saving ``Redirect`` objects.
+    """
+
+    model = Redirect
+
+    def initialize_instance(self, kwargs):
+        # Automatically normalise paths when creating instances from kwargs
+        kwargs["old_path"] = Redirect.normalise_path(kwargs["old_path"])
+        return super().initialize_instance(kwargs)
+
+    def pre_process(self):
+        # delete any existing automatically-created redirects that might clash
+        # with the items in `self.items`
+        clashes_q = Q()
+        for item in self.items:
+            clashes_q |= Q(old_path=item.old_path, site_id=item.site_id)
+        Redirect.objects.filter(automatically_created=True).filter(clashes_q).delete()
+
+
+def autocreate_redirects(
+    sender: PageBase, instance: Page, url_path_before: str, url_path_after: str, log_entry: Optional[BaseLogEntry] = None
+) -> None:
+    if instance.is_site_root():
+        # url_path changes for site root pages do not affect URLs generally,
+        # so we can exit early here.
+        return None
+
+    # This shouldn't ever be the case, but...
+    assert url_path_before != url_path_after
+
+    logger.info(f"Creating redirects for page: '{instance}' id={instance.id}")
+    batch = BatchRedirectCreator(
+        max_size=getattr(settings, "WAGTAILREDIRECTS_AUTOCREATE_BATCH_SIZE", 2000),
+        ignore_conflicts=True,
+    )
+    request = get_dummy_request()
+
+    for page in (
+        instance.get_descendants(inclusive=True)
+        .ever_live()
```

Comment:
Why do we need to create redirects for pages that have been unpublished?

Reply:
I was thinking we might have pages that have been unpublished temporarily or something - but, now I'm questioning the usefulness of that.

Reply:
I think in my head there's a difference between a page that has been unpublished and might be published again, to a page that has been deleted. I think if I was to unpublish a page for a short time, it's still the same page from a BE perspective, and so I think it makes sense for it to be considered the same page from the FE perspective too... so - someone visiting the old URL is taken to the new version when it is published again. Does that make sense to you?

Reply:
I'm wondering if we should instead create the redirect when the page is unpublished? It's similar to deleting it I think

Reply:
I suppose it might be nice if the UI asked "Do you want to send visitors elsewhere?" when unpublishing - like we do for deleting. With that, I would certainly not bother creating redirects for unpublished pages here. Tell you what - I'll create issues for that functionality, and reference this conversation from there. Sound good?

Reply:
Okay, issues created:
- #7790 
- #7791